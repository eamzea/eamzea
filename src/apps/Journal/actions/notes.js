import { db } from "../firebase/firebaseConfig";
import { types } from "../types/types";
import { loadNotes } from "../helpers/loadNotes";
import Swal from "sweetalert2";
import { fileUpload } from "../helpers/fileUpload";

export const startNewNote = () => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;

    const newNote = {
      title: "",
      body: "",
      date: new Date().getTime(),
    };

    const doc = await db.collection(`${uid}/journal/notes`).add(newNote);

    dispatch(activeNote(doc.id, newNote));
    dispatch(addNewNote(doc.id, newNote));
  };
};

export const activeNote = (id, note) => ({
  type: types.notesNoteActive,
  payload: {
    id,
    ...note,
  },
});

export const addNewNote = (id, note) => ({
  type: types.notesAddNew,
  payload: {
    id,
    ...note,
  },
});

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  };
};

export const setNotes = (notes) => ({
  type: types.notesLoad,
  payload: notes,
});

export const startSaveNote = (note) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;

    if (!note.url) {
      note.url =
        "https://images.immediate.co.uk/production/volatile/sites/4/2018/08/untitled-0888e07.jpg?quality=90&resize=940%2C400";
    }

    const noteToFirestore = { ...note };

    delete noteToFirestore.id;

    await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);

    dispatch(refreshNote(note.id, noteToFirestore));
    Swal.fire("Saved", note.title, "success");
  };
};

export const refreshNote = (id, note) => ({
  type: types.notesUpdated,
  payload: {
    id,
    note: {
      id,
      ...note,
    },
  },
});

export const startUploading = (file) => {
  return async (dispatch, getState) => {
    const activeNote = getState().notes.active;

    Swal.fire({
      title: "Uploading...",
      text: "Please wait...",
      allowOutsideClick: false,
      onBeforeOpen: () => {
        Swal.showLoading();
      },
    });

    const fileUrl = await fileUpload(file);
    activeNote.url = fileUrl;

    dispatch(startSaveNote(activeNote));

    Swal.close();
  };
};

export const startDeleting = (id) => {
  return async (dispatch, getState) => {
    const result = await Swal.fire({
      title: "Do you want to delete this note?",
      showDenyButton: true,
      confirmButtonText: `Yes`,
      denyButtonText: `No`,
      allowOutsideClick: false,
    });

    if (result.isConfirmed) {
      const uid = getState().auth.uid;
      await db.doc(`${uid}/journal/notes/${id}`).delete();

      dispatch(deleteNote(id));

      Swal.fire("Note has been deleted!", "", "success");
    } else if (result.isDenied) {
      Swal.fire(`Note hasn't been deleted`, "", "info");
    }
  };
};

export const deleteNote = (id) => ({
  type: types.notesDelete,
  payload: id,
});

export const notesLogout = () => ({
  type: types.notesLogoutCleaning,
});
