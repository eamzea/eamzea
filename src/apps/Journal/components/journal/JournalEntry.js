import React from "react";
import moment from "moment";
import { useDispatch } from "react-redux";
import { activeNote } from "../../actions/notes";

const JournalEntry = ({ id, title, body, date, url }) => {
  const dispatch = useDispatch();

  const noteDay = moment(date);

  const handleEntryClick = () => {
    dispatch(
      activeNote(id, {
        date,
        title,
        body,
        url,
      })
    );
  };

  return (
    <div
      className="journal__entry pointer animate__animated animate__faster animate__slideInLeft"
      onClick={handleEntryClick}
    >
      <div
        className="journal__entry-picture"
        style={{
          backgroundSize: "cover",
          backgroundPosition: "center",
          backgroundImage: url
            ? `url(${url})`
            : "url(https://i.pinimg.com/originals/64/29/75/642975b5a9f8b8414d5f3d4a6415fefc.jpg)",
        }}
      ></div>
      <div className="journal__entry-body">
        <p className="journal__entry-title">{title || "Without Title"}</p>
        <p className="journal__entry-content">{body || "Without Body"}</p>
      </div>
      <div className="journal__entry-date-box">
        <span>{noteDay.format("ddd")}</span>
        <h4>{noteDay.format("Do")}</h4>
      </div>
    </div>
  );
};

export default JournalEntry;
