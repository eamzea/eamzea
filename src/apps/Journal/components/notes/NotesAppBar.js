import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { startSaveNote, startUploading } from "../../actions/notes";
import moment from "moment";

const NotesAppBar = () => {
  const dispatch = useDispatch();

  const { active: note } = useSelector((state) => state.notes);

  const handleSave = () => {
    dispatch(startSaveNote(note));
  };

  const handlePictureUpload = () => {
    document.querySelector("#fileSelector").click();
  };

  const handleFileChange = (e) => {
    const file = e.target.files[0];

    if (file) {
      dispatch(startUploading(file));
    }
  };

  const today = moment().format("dddd, MMMM Do, YYYY");

  return (
    <div className="notes__appbar">
      <span>{today}</span>

      <input
        id="fileSelector"
        name="file"
        type="file"
        style={{ display: "none" }}
        onChange={handleFileChange}
      />

      <div>
        <button className="btn" onClick={handlePictureUpload}>
          Picture
        </button>
        <button className="btn" onClick={handleSave}>
          Save
        </button>
      </div>
    </div>
  );
};

export default NotesAppBar;
