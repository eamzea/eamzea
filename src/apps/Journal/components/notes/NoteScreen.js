import React, { useEffect, useRef } from "react";
import NotesAppBar from "./NotesAppBar";
import { useSelector, useDispatch } from "react-redux";
import { useForm } from "../../hooks/useForm";
import { activeNote, startDeleting } from "../../actions/notes";

const NoteScreen = () => {
  const dispatch = useDispatch();

  const { active: note } = useSelector((state) => state.notes);

  const [formValues, handleInputChange, reset] = useForm(note);

  const { title, body, id } = formValues;

  const activeId = useRef(note.id);

  useEffect(() => {
    if (note.id !== activeId.current) {
      reset(note);
      activeId.current = note.id;
    }
  }, [note, reset]);

  useEffect(() => {
    dispatch(activeNote(formValues.id, { ...formValues }));
  }, [formValues, dispatch]);

  const handleDelete = () => {
    dispatch(startDeleting(id));
  };

  return (
    <div className="notes__main-content">
      <NotesAppBar />
      <div className="notes__content">
        <input
          type="text"
          name="title"
          placeholder="Some awesome title"
          className="notes__input-title"
          autoComplete="off"
          value={title}
          onChange={handleInputChange}
        />
        <textarea
          name="body"
          placeholder="What happened today?"
          className="notes__textarea"
          value={body}
          onChange={handleInputChange}
        ></textarea>
      </div>
      <div className="notes__footer-container">
        <div
          className="notes__image"
          style={{
            backgroundImage: note.url
              ? `url(${note.url})`
              : "url(https://images.immediate.co.uk/production/volatile/sites/4/2018/08/untitled-0888e07.jpg?quality=90&resize=940%2C400)",
          }}
        ></div>
        <div className="notes__container-button">
          <button
            className="btn btn-danger"
            style={{ width: "50%", margin: "5px 0px" }}
            onClick={handleDelete}
          >
            <i className="fas fa-trash"></i>
          </button>
        </div>
      </div>
    </div>
  );
};

export default NoteScreen;
