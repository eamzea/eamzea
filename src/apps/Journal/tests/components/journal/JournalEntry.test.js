import React from "react";
import { mount } from "enzyme";
import { Provider } from "react-redux";

import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import "@testing-library/jest-dom";
import JournalEntry from "../../../components/journal/JournalEntry";
import { activeNote } from "../../../actions/notes";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};

const note = {
  id: 10,
  date: 0,
  title: "Note Test",
  body: "Body Test",
  url: "https://randomimages.com/image.png",
};

let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <JournalEntry {...note} />
  </Provider>
);

describe("Pruebas en <JournalEntry/>", () => {
  test("Debe mostrarse correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe disparar el handleClick", () => {
    wrapper.find(".journal__entry").prop("onClick")();

    expect(store.dispatch).toHaveBeenCalledWith(
      activeNote(note.id, { ...note })
    );
  });
});
