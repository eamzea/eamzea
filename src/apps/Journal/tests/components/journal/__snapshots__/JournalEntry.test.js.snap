// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <JournalEntry/> Debe mostrarse correctamente 1`] = `
<div
  className="journal__entry pointer animate__animated animate__faster animate__slideInLeft"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://randomimages.com/image.png)",
        "backgroundPosition": "center",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Note Test
    </p>
    <p
      className="journal__entry-content"
    >
      Body Test
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Wed
    </span>
    <h4>
      31st
    </h4>
  </div>
</div>
`;
